src/mame/drivers/marblmd2.cpp

```diff
@@ -0,0 +1,388 @@
+{
+	// start drawing
+	m_vad->mob().draw_async(cliprect);
+
+	// draw the playfield
+	bitmap_ind8 &priority_bitmap = screen.priority();
+	priority_bitmap.fill(0, cliprect);
+	m_vad->playfield().draw(screen, m_tempbitmap, cliprect, 0, 0x00);
+	bitmap_ind16 &mobitmap = m_vad->mob().bitmap();
+
+	for (int y = cliprect.min_y; y <= cliprect.max_y; y++)
+	{
+		uint16_t *const src = &m_tempbitmap.pix(y);
+		uint16_t *const dst = &bitmap.pix(y);
+
+		// top bit of the gfxdata appears to be priority, so we don't want it in the render bitmap
+		for (int x = cliprect.min_x; x <= cliprect.max_x; x++)
+		{
+			dst[x] = src[x] & 0x007f;
+		}
+	}
+
+	for (const sparse_dirty_rect *rect = m_vad->mob().first_dirty_rect(cliprect); rect != nullptr; rect = rect->next())
+	{
+		for (int y = rect->top(); y <= rect->bottom(); y++)
+		{
+			uint16_t *const pf2 = &m_tempbitmap.pix(y);
+
+			uint16_t const *const mo = &mobitmap.pix(y);
+			uint16_t *const pf = &bitmap.pix(y);
+			for (int x = rect->left(); x <= rect->right(); x++)
+			{
+				if (mo[x] != 0xffff)
+				{
+					if (pf2[x] & 0x80) // check against top bit of temp pf render
+					{
+						if ((mo[x] & 0x80) == 0x80)
+							pf[x] = mo[x];
+					}
+					else
+					{
+						pf[x] = mo[x] | 0x80;
+					}
+
+				}
+			}
+		}
+	}
+
+
+	return 0;
+}
+
+// not verified
+void marblmd2_state::latch_w(offs_t offset, uint16_t data, uint16_t mem_mask)
+{
+	COMBINE_DATA(&m_latch_data);
+
+	if (m_latch_data & 0x0010)
+		m_jsa->soundcpu().set_input_line(INPUT_LINE_RESET, CLEAR_LINE);
+	else
+		m_jsa->soundcpu().set_input_line(INPUT_LINE_RESET, ASSERT_LINE);
+}
+
+
+void marblmd2_state::marblmd2_map(address_map& map)
+{
+	map(0x000000, 0x07ffff).rom();
+
+	map(0x600000, 0x600001).portr("600000");
+	map(0x600002, 0x600003).portr("600002");
+	map(0x600010, 0x600011).portr("600010");
+	map(0x600012, 0x600013).portr("600012");
+	map(0x600020, 0x600021).portr("600020");
+	map(0x600030, 0x600031).r(m_jsa, FUNC(atari_jsa_iii_device::main_response_r));
+	map(0x600040, 0x600041).w(m_jsa, FUNC(atari_jsa_iii_device::main_command_w));
+
+	map(0x600050, 0x600051).w(FUNC(marblmd2_state::latch_w));
+
+	map(0x601000, 0x6013ff).ram(); // some kind of NVRAM?
+
+	map(0x607000, 0x607001).nopw();
+	
+	map(0x7c0000, 0x7c03ff).rw("palette", FUNC(palette_device::read8), FUNC(palette_device::write8)).umask16(0xff00).share("palette");
+	map(0x7cffc0, 0x7cffff).ram().rw(m_vad, FUNC(atari_vad_device::control_read), FUNC(atari_vad_device::control_write));
+
+	map(0x7d0000, 0x7d7fff).ram();
+	map(0x7d8000, 0x7d9fff).ram().w(m_vad, FUNC(atari_vad_device::playfield_latched_lsb_w)).share("vad:playfield");
+	map(0x7da000, 0x7dbeff).ram().share("vad:mob");
+	map(0x7dbf00, 0x7dbf7f).ram().share("vad:eof");
+	map(0x7dbf80, 0x7dbfff).ram().share("vad:mob:slip");
+	map(0x7f8000, 0x7fbfff).ram();
+}
+
+static INPUT_PORTS_START( marblmd2 )
+
+	PORT_START("600000")
+	PORT_BIT( 0x00ff, IP_ACTIVE_LOW, IPT_UNKNOWN )
+	PORT_BIT( 0x0100, IP_ACTIVE_LOW, IPT_START1 )
```

Comment:
These start buttons, do they double function as action buttons? (It's common with Atari games). If so, BUTTON1 is a better choice. Where is player 3's button?

Reply:
They do, and it looks like I lost the P3 button in a code tidy earlier